FreeRTOS/Demo/ARC_NSIM_QEMU_GCC_MW/arc/arc_exception.c

```diff
@@ -0,0 +1,260 @@
+/* ------------------------------------------
+ * Copyright (c) 2022, Synopsys, Inc. All rights reserved.
+
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+
+ * 1) Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+
+ * 2) Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation and/or
+ * other materials provided with the distribution.
+
+ * 3) Neither the name of the Synopsys, Inc., nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+   --------------------------------------------- */
+
+/**
+ * @file
+ * @ingroup ARC_HAL_EXCEPTION_CPU ARC_HAL_EXCEPTION_INTERRUPT
+ * @brief C Implementation of exception and interrupt management
+ */
+#include "arc_exception.h"
+
+/**
+ * @addtogroup ARC_HAL_EXCEPTION_CPU
+ * @{
+ * @var exc_entry_table
+ * @brief exception entry table
+ *
+ * install exception entry table to ARC_AUX_INT_VECT_BASE in startup.
+ * According to ARCv2 ISA, vectors are fetched in instruction space and thus
+ * may be present in ICCM, Instruction Cache, or
+ * main memory accessed by instruction fetch logic.
+ * So it is put into a specific section .vector.
+ *
+ * Please note that the exc_entry_table maybe cached in ARC. Some functions is
+ * defined in .s files.
+ *
+ */
+
+/**
+ * @ingroup ARC_HAL_EXCEPTION_CPU
+ * @brief  default cpu exception handler
+ * @param p_excinf pointer to the exception frame
+ */
+static void exc_handler_default(void *p_excinf)
+{
+	uint32_t excpt_cause_reg = 0;
+	uint32_t excpt_ret_reg = 0;
+	uint32_t exc_no = 0;
+	uint32_t exc_cause = 0;
+	uint32_t exc_param = 0;
+
+	excpt_cause_reg = arc_aux_read(AUX_ECR);
+	excpt_ret_reg = arc_aux_read(AUX_ERRET);
+	exc_no = (excpt_cause_reg >> 16) & 0xff;
+	exc_cause = (excpt_cause_reg >> 8) & 0xff;
+	exc_param = (excpt_cause_reg >> 0) & 0xff;
```

Comment:
I'm wondering why do we have this?

As we don't use result the compiler is free to optimize this out.